### PR TITLE
compression: add generic decompressor filter

### PR DESCRIPTION
**Pull request by junr03**
Commit Message: add generic decompressor filter
Additional Description: this PR takes the design of the compressor filter, and uses the decompressor extension point to build a generic decompressor http filter. This filter can be used to decompress compressed payloads both on the request and response direction.
Risk Level: low - low as it is an extension, med - for users as this is a brand new filter.
Testing: unit tests, integration tests
Docs Changes: added docs
Release Notes: added release notes

### CODE REVIEW

**Comment by dio** [comment]
cc. @rgs1 probably you're interested in this 🙂.

File: source/extensions/filters/http/decompressor/decompressor_filter.cc [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    direction_config.stats().decompressed_.inc();
+    decompressor = config_->makeDecompressor();
+    // remove the compression scheme's content encoding because the filter is going to decompress.
+    removeContentEncoding(headers);
+
+    // Per the filter documentation, only buffer if there is a Content-Length header present, and
+    // the filter is configured to buffer.
+    if (headers.removeContentLength() > 0 && direction_config.maxBufferedBytes()) {
+      direction_config.setBuffering(true);
+      callbacks.setBufferLimit(direction_config.maxBufferedBytes().value());
```
Comment (junr03) [inline source/extensions/filters/http/decompressor/decompressor_filter.cc:115]: @alyssawilk I have this logic that acts on streams regardless of request/response. I need to interact with the stream callbacks and I was wondering why we don't have a virtual `setBufferLimit` function instead (or in addition to) the `setDecoder/EncoderBufferLimit` in the derived classes?

Comment (alyssawilk) [reply source/extensions/filters/http/decompressor/decompressor_filter.cc:115]: If you look at what's in StreamFilterCallbacks it affects everything about the stream.  There's one stream ID.  If you reset the stream it is reset.  if you SetBufferLimits on the stream and it only sets the request or response buffer limits IMO it's confusing.    In general it's better to have the clarity on what limits you're setting so folks can tell if they're affecting the request or response path even if it involves some down casts or absl::variants, but that's just my opinion.

Comment (junr03) [reply source/extensions/filters/http/decompressor/decompressor_filter.cc:115]: Thanks for the explanation! Makes sense, all functionality in the base class affects the stream as a whole, not individual directions. I'll rework this.

**Comment by junr03** [comment]
@rojkov @rebello95 I have addressed all the comments and thoughts we had discussed in the rough draft of this PR. I have also added a bunch of unit tests.

I am only missing integration tests. But if you both could give this another look I would appreciate it.

File: source/extensions/filters/http/decompressor/decompressor_filter.cc [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+      }
+      direction_config.setHeaders(&headers);
+      // Note that the log will print the updated headers. The incoming headers can be seen from the
+      // log in decode/encodeHeaders.
+      ENVOY_STREAM_LOG(debug, "do decompress (without buffering) {}: {}", callbacks,
+                       direction_config.logString(), headers);
+      return Http::FilterHeadersStatus::StopIteration;
+    }
+    // FIX ME: chunked should not be added if we buffer, right?
+    sanitizeTransferEncoding(headers);
```
Comment (rojkov) [inline source/extensions/filters/http/decompressor/decompressor_filter.cc:132]: That is my understanding too: if `chunked` is added then `Content-Length` has to be omitted.

File: source/extensions/filters/http/decompressor/decompressor_filter.cc [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus
+DecompressorFilter::maybeDecompress(DecompressorFilterConfig::DirectionConfig& direction_config,
+                                    Compression::Decompressor::Decompressor* decompressor,
+                                    Http::StreamFilterCallbacks& callbacks,
+                                    Buffer::Instance& input_buffer, const bool end_stream) const {
+  if (decompressor) {
```
Comment (rojkov) [inline source/extensions/filters/http/decompressor/decompressor_filter.cc:150]: I'd rather make `decompressor` an unique pointer and compare it with `nullptr` here.

File: api/envoy/extensions/filters/http/decompressor/v3/decompressor.proto [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+  // Configuration for filter behavior on the response direction.
+  message ResponseDirectionConfig {
+    CommonDirectionConfig common_config = 1;
+  }
+
+  // A decompressor library to use for decompression. Currently only
+  // :ref:`envoy.compression.gzip.compressor<envoy_api_msg_extensions.compression.gzip.decompressor.v3.Gzip>`
+  // is included in Envoy.
+  config.core.v3.TypedExtensionConfig decompressor_library = 1
```
Comment (rojkov) [inline api/envoy/extensions/filters/http/decompressor/v3/decompressor.proto:57]: If we configure bufferization separately for requests and responses shouldn't we have separate configs for the decompressor library as well?

File: source/extensions/filters/http/decompressor/decompressor_filter.cc (hunk 3 above)
Comment (junr03) [reply source/extensions/filters/http/decompressor/decompressor_filter.cc:150]: I could make it an optional ref if you'd rather not pass the raw pointer. However, I don't really want to move the ownership of the pointer into the function.

File: api/envoy/extensions/filters/http/decompressor/v3/decompressor.proto (hunk 4 above)
Comment (junr03) [reply api/envoy/extensions/filters/http/decompressor/v3/decompressor.proto:57]: We _could_ do that. And the stats would work as well.... I am not sure if that is increasing the cognitive load of someone trying to configure more.

**Comment by junr03** [comment]
@rojkov while talking with @rebello95 today I had an idea: instead of doing buffering in the decompressor filter to preserve the content length header, people can install the buffer filter after decompression to get the same outcome. This seems preferable to me than making the decompressor filter more complex so I removed buffering here https://github.com/envoyproxy/envoy/pull/11172/commits/42de78bc4ee747b140905d2a59bf873fa971f124. 

wdyt?

File: source/extensions/filters/http/decompressor/decompressor_filter.cc [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+};
+
+Http::FilterDataStatus DecompressorFilter::decodeData(Buffer::Instance& data, bool) {
+  return maybeDecompress(
+      config_->requestDirectionConfig(),
+      request_decompressor_
+          ? Envoy::Compression::Decompressor::DecompressorOptRef(std::ref(*request_decompressor_))
+          : absl::nullopt,
+      *decoder_callbacks_, data);
+}
```
Comment (rojkov) [inline source/extensions/filters/http/decompressor/decompressor_filter.cc:81]: Would it be simpler to have here something like this instead?
```c++
Http::FilterDataStatus DecompressorFilter::encodeData(Buffer::Instance& data, bool) {
    if (response_decompressor_ != nullptr) then {
        doDecompress(config_->requestDirectionConfig(), request_decompressor_, *decoder_callbacks_, data);
    }

    return  Http::FilterDataStatus::Continue;
}
```
Proper ownership and memory cleanup would be still there if `doCompress()` accepted a const reference to the unique pointer like in `maybeInitDescompress()`.

File: source/extensions/filters/http/decompressor/decompressor_filter.cc [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+
+Http::FilterDataStatus DecompressorFilter::encodeData(Buffer::Instance& data, bool) {
+  return maybeDecompress(
+      config_->requestDirectionConfig(),
+      response_decompressor_
+          ? Envoy::Compression::Decompressor::DecompressorOptRef(std::ref(*response_decompressor_))
+          : absl::nullopt,
+      *decoder_callbacks_, data);
+}
```
Comment (rojkov) [inline source/extensions/filters/http/decompressor/decompressor_filter.cc:102]: same here

File: source/extensions/filters/http/decompressor/decompressor_filter.cc [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+  if (end_stream) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+  ENVOY_STREAM_LOG(debug, "DecompressorFilter::decodeHeaders: {}", *decoder_callbacks_, headers);
+
+  // Two responsibilities on the request side:
+  //   1. If response decompression is enabled (and advertisement is enabled), then advertise to
+  //      the upstream that this hop is able to decompress responses via the Accept-Encoding header.
+  if (config_->responseDirectionConfig().decompressionEnabled()) {
+    if (config_->requestDirectionConfig().advertiseAcceptEncoding()) {
```
Comment (rebello95) [inline source/extensions/filters/http/decompressor/decompressor_filter.cc:61]: nit: You could reduce a level of nesting here by combining these

File: source/extensions/filters/http/decompressor/decompressor_filter.cc [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+    input_buffer.add(output_buffer);
+  }
+  return Http::FilterDataStatus::Continue;
+}
+
+bool DecompressorFilter::hasCacheControlNoTransform(
+    Http::RequestOrResponseHeaderMap& headers) const {
+  return headers.CacheControl()
+             ? StringUtil::caseFindToken(headers.CacheControl()->value().getStringView(), ",",
+                                         Http::Headers::get().CacheControlValues.NoTransform)
```
Comment (rebello95) [inline source/extensions/filters/http/decompressor/decompressor_filter.cc:153]: Should this be case insensitive?

File: source/extensions/filters/http/decompressor/decompressor_filter.cc [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+             : false;
+}
+
+bool DecompressorFilter::contentEncodingMatches(Http::RequestOrResponseHeaderMap& headers) const {
+  if (headers.ContentEncoding()) {
+    absl::string_view coding = StringUtil::trim(
+        StringUtil::cropRight(headers.ContentEncoding()->value().getStringView(), ","));
+    if (StringUtil::CaseInsensitiveCompare()(config_->contentEncoding(), coding)) {
+      return true;
+    }
```
Comment (rebello95) [inline source/extensions/filters/http/decompressor/decompressor_filter.cc:163]:
```suggestion
    return StringUtil::CaseInsensitiveCompare()(config_->contentEncoding(), coding);
```

File: source/extensions/filters/http/decompressor/decompressor_filter.cc [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+
+void DecompressorFilter::removeContentEncoding(Http::RequestOrResponseHeaderMap& headers) const {
+  const auto all_codings = headers.ContentEncoding()->value().getStringView();
+  const auto remaining_codings = StringUtil::trim(StringUtil::cropLeft(all_codings, ","));
+
+  if (remaining_codings != all_codings) {
+    headers.setContentEncoding(remaining_codings);
+  } else {
+    headers.removeContentEncoding();
+  }
```
Comment (rebello95) [inline source/extensions/filters/http/decompressor/decompressor_filter.cc:176]: It might be clearer here to check if `remaining_codings` is empty instead to convey intent

File: source/extensions/filters/http/decompressor/decompressor_filter.cc [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+
+/**
+ * Add "chunked" to the Transfer-Encoding header if it's not there yet.
+ */
+void DecompressorFilter::sanitizeTransferEncoding(Http::RequestOrResponseHeaderMap& headers) const {
+  const Http::HeaderEntry* transfer_encoding = headers.TransferEncoding();
+  if (!(transfer_encoding &&
+        StringUtil::caseFindToken(transfer_encoding->value().getStringView(), ",",
+                                  Http::Headers::get().TransferEncodingValues.Chunked,
+                                  true /* trim_whitespace */))) {
```
Comment (rebello95) [inline source/extensions/filters/http/decompressor/decompressor_filter.cc:187]: Do we need to trim whitespace elsewhere in the above header checks too?

File: test/extensions/filters/http/decompressor/decompressor_filter_test.cc [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+  EXPECT_CALL(*decompressor_factory_, createDecompressor())
+      .WillOnce(Return(ByMove(std::move(decompressor))));
+  Http::TestHeaderMapImpl headers_before_filter{{"content-encoding", "mock"},
+                                                {"content-length", "256"}};
+  std::unique_ptr<Http::RequestOrResponseHeaderMap> headers_after_filter =
+      doHeaders(headers_before_filter, false /* end_stream */);
+
+  EXPECT_EQ(nullptr, headers_after_filter->ContentEncoding());
+
+  // FIX ME: pending decision on this.
```
Comment (rebello95) [inline test/extensions/filters/http/decompressor/decompressor_filter_test.cc:126]: We've decided to remove content length, right?

File: test/extensions/filters/http/decompressor/decompressor_filter_test.cc [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+  EXPECT_CALL(*decompressor_factory_, createDecompressor())
+      .WillOnce(Return(ByMove(std::move(decompressor))));
+  Http::TestHeaderMapImpl headers_before_filter{{"content-encoding", "mock, br"},
+                                                {"content-length", "256"}};
+  std::unique_ptr<Http::RequestOrResponseHeaderMap> headers_after_filter =
+      doHeaders(headers_before_filter, false /* end_stream */);
+
+  EXPECT_EQ("br", headers_after_filter->ContentEncoding()->value().getStringView());
+
+  // FIX ME: pending decision on this.
```
Comment (rebello95) [inline test/extensions/filters/http/decompressor/decompressor_filter_test.cc:152]: Same here

File: test/extensions/filters/http/decompressor/decompressor_filter_test.cc [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+      .WillOnce(Return(ByMove(std::move(decompressor))));
+  Http::TestHeaderMapImpl headers_before_filter{{"content-encoding", "mock, br"},
+                                                {"content-length", "256"},
+                                                {"transfer-encoding", "chunked"}};
+  std::unique_ptr<Http::RequestOrResponseHeaderMap> headers_after_filter =
+      doHeaders(headers_before_filter, false /* end_stream */);
+
+  EXPECT_EQ("br", headers_after_filter->ContentEncoding()->value().getStringView());
+
+  // FIX ME: pending decision on this.
```
Comment (rebello95) [inline test/extensions/filters/http/decompressor/decompressor_filter_test.cc:173]: Same here

File: test/extensions/filters/http/decompressor/decompressor_filter_test.cc [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+
+TEST_P(DecompressorFilterTest, DecompressionActiveTransferEncodingPresentAlready) {
+  // Keep the decompressor to set expectations about it
+  auto decompressor = std::make_unique<Compression::Decompressor::MockDecompressor>();
+  auto* decompressor_ptr = decompressor.get();
+  EXPECT_CALL(*decompressor_factory_, createDecompressor())
+      .WillOnce(Return(ByMove(std::move(decompressor))));
+  Http::TestHeaderMapImpl headers_before_filter{{"content-encoding", "mock, br"},
+                                                {"content-length", "256"},
+                                                {"transfer-encoding", "chunked"}};
```
Comment (rebello95) [inline test/extensions/filters/http/decompressor/decompressor_filter_test.cc:167]: Should we set this to something custom like `mock` instead to validate that it isn't the filter setting `chunked` (like we're validating in the previous test)?

File: test/extensions/filters/http/decompressor/decompressor_filter_test.cc [hunk 16]
```diff
@@ -0,0 +1,281 @@
...
+  Http::TestHeaderMapImpl headers_before_filter{{"content-encoding", "not-matching"},
+                                                {"content-length", "256"}};
+  std::unique_ptr<Http::RequestOrResponseHeaderMap> headers_after_filter =
+      doHeaders(headers_before_filter, false /* end_stream */);
+  TestUtility::headerMapEqualIgnoreOrder(headers_before_filter, *headers_after_filter);
+
+  expectNoDecompression();
+}
+
+TEST_P(DecompressorFilterTest, NoDecompressionContentEncodingNotCurrent) {
```
Comment (rebello95) [inline test/extensions/filters/http/decompressor/decompressor_filter_test.cc:226]: What does "not current" mean in this context?

File: test/extensions/filters/http/decompressor/decompressor_filter_test.cc [hunk 17]
```diff
@@ -0,0 +1,281 @@
...
+      {"content-encoding", "mock"},
+      {"content-length", "256"}};
+  std::unique_ptr<Http::RequestOrResponseHeaderMap> headers_after_filter =
+      doHeaders(headers_before_filter, false /* end_stream */);
+  TestUtility::headerMapEqualIgnoreOrder(headers_before_filter, *headers_after_filter);
+
+  expectNoDecompression();
+}
+
+TEST_P(DecompressorFilterTest, DisableAdvertiseAcceptEncoding) {
```
Comment (rebello95) [inline test/extensions/filters/http/decompressor/decompressor_filter_test.cc:250]: Should we also test the `EnableAdvertiseAcceptEncoding` case?

File: test/extensions/filters/http/decompressor/decompressor_filter_test.cc [hunk 18]
```diff
@@ -0,0 +1,281 @@
...
+      doHeaders(headers_before_filter, false /* end_stream */);
+
+  EXPECT_EQ(nullptr, headers_after_filter->ContentEncoding());
+  EXPECT_EQ(nullptr, headers_after_filter->ContentLength());
+  EXPECT_EQ("chunked", headers_after_filter->TransferEncoding()->value().getStringView());
+  EXPECT_EQ(nullptr, headers_after_filter->get(Http::LowerCaseString{"accept-encoding"}));
+
+  expectDecompression(decompressor_ptr);
+}
+
```
Comment (rebello95) [inline test/extensions/filters/http/decompressor/decompressor_filter_test.cc:276]: I think it might make sense to battle test some of the header parsing logic a bit more as well (i.e., case insensitive headers, spacing, etc.)

File: source/extensions/filters/http/decompressor/decompressor_filter.cc (hunk 10 above)
Comment (junr03) [reply source/extensions/filters/http/decompressor/decompressor_filter.cc:176]: The way this is working if we are removing the last coding, the remaining_codings won't be empty.

File: source/extensions/filters/http/decompressor/decompressor_filter.cc (hunk 11 above)
Comment (junr03) [reply source/extensions/filters/http/decompressor/decompressor_filter.cc:187]: which one, in particular?

Comment (rebello95) [reply source/extensions/filters/http/decompressor/decompressor_filter.cc:187]: `contentEncodingMatches` and `hasCacheControlNoTransform`

File: test/extensions/filters/http/decompressor/decompressor_filter_test.cc (hunk 15 above)
Comment (junr03) [reply test/extensions/filters/http/decompressor/decompressor_filter_test.cc:167]: I have actually decided to remove the transfer encoding logic. From my reading of documentation it actually doesn't seem that we should be setting it. AFAICT if the data were actually chunked we would have to transform the data chunks [per](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding):

> at the beginning of each chunk you need to add the length of the current chunk in hexadecimal format, followed by '\r\n' and then the chunk itself, followed by another '\r\n'. The terminating chunk is a regular chunk, with the exception that its length is zero. It is followed by the trailer, which consists of a (possibly empty) sequence of entity header fields. 

which this filter is not doing. Also this is a http/1.1 construct.

@rojkov am I missing something here?

File: source/extensions/filters/http/decompressor/decompressor_filter.cc (hunk 8 above)
Comment (junr03) [reply source/extensions/filters/http/decompressor/decompressor_filter.cc:153]: caseFindToken is case insensitive.

File: test/extensions/filters/http/decompressor/decompressor_filter_test.cc (hunk 18 above)
Comment (junr03) [reply test/extensions/filters/http/decompressor/decompressor_filter_test.cc:302]: added more tests

File: source/extensions/filters/http/decompressor/decompressor_filter.cc (hunk 11 above)
Comment (junr03) [reply source/extensions/filters/http/decompressor/decompressor_filter.cc]: they do trimming were necessary.

File: api/envoy/extensions/filters/http/decompressor/v3/decompressor.proto (hunk 4 above)
Comment (junr03) [reply api/envoy/extensions/filters/http/decompressor/v3/decompressor.proto:49]: I think I'll leave as is to future proof in case people want to add more complexity.

**Comment by junr03** [comment]
@rojkov, @rebello95 updated if you can take a look

File: test/extensions/filters/http/decompressor/decompressor_filter_test.cc (hunk 15 above)
Comment (rojkov) [reply test/extensions/filters/http/decompressor/decompressor_filter_test.cc]: AFAIU parsing and forming data chunks gets done in the http1 codec, not in filters. It might be that the codec sets the header itself if needed, but I'm not sure. Better to checks the codec's sources.

File: source/extensions/filters/http/decompressor/decompressor_filter.h [hunk 19]
```diff
@@ -134,15 +134,15 @@ class DecompressorFilter : public Http::PassThroughFilter,
...
                   Http::StreamFilterCallbacks& callbacks, Buffer::Instance& input_buffer) const;
 
   // TODO(junr03): these do not need to be member functions. They can all be part of a static
   // utility class. Moreover, they can be shared between compressor and decompressor.
   bool hasCacheControlNoTransform(Http::RequestOrResponseHeaderMap& headers) const;
   bool contentEncodingMatches(Http::RequestOrResponseHeaderMap& headers) const;
-  void removeContentEncoding(Http::RequestOrResponseHeaderMap& headers) const;
-  void sanitizeTransferEncoding(Http::RequestOrResponseHeaderMap& headers) const;
+  void modifyContentEncoding(Http::RequestOrResponseHeaderMap& headers) const;
+  void modifyTransferEncoding(Http::RequestOrResponseHeaderMap& headers) const;
```
Comment (rebello95) [inline source/extensions/filters/http/decompressor/decompressor_filter.h:145]: Looks like this is unused now?

File: test/extensions/filters/http/decompressor/decompressor_filter_test.cc [hunk 20]
```diff
@@ -111,70 +145,73 @@ INSTANTIATE_TEST_SUITE_P(IsRequestDirection, DecompressorFilterTest,
...
+  // Additional spacing should still match.
+  Http::TestHeaderMapImpl headers_before_filter{{"content-encoding", " mock "},
+                                                {"content-length", "256"}};
+  decompressionActive(headers_before_filter, absl::nullopt /* expected_content_encoding */);
+}
 
-  expectDecompression(decompressor_ptr);
+TEST_P(DecompressorFilterTest, DecompressionActiveContentEncodingCasing) {
+  // Different casing should still match.
+  Http::TestHeaderMapImpl headers_before_filter{{"content-encoding", "MOCK"},
```
Comment (rebello95) [inline test/extensions/filters/http/decompressor/decompressor_filter_test.cc:162]: Are keys forced to be lowercased at the point they're passed to filters? If not, we should probably handle that as well (though I assume they are sanitized correctly)

Comment (junr03) [reply test/extensions/filters/http/decompressor/decompressor_filter_test.cc:162]: yep, this is all avoided because keys to headers are actually instances of the LowerCaseString class. This is just a convenience constructor.

File: test/extensions/filters/http/decompressor/decompressor_filter_test.cc (hunk 15 above)
Comment (junr03) [reply test/extensions/filters/http/decompressor/decompressor_filter_test.cc]: yep the http/1.1 codec handles this https://github.com/envoyproxy/envoy/blob/e3f1506a0c6bbe7afa0980cca58611e105f3e350/source/common/http/http1/codec_impl.cc#L139

**Comment by junr03** [comment]
@mattklein123 this is now ready for you! Thanks

File: api/envoy/extensions/filters/http/decompressor/v3/decompressor.proto [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+      [(validate.rules).message = {required: true}];
+
+  RequestDirectionConfig request_direction_config = 2;
+
+  ResponseDirectionConfig response_direction_config = 3;
```
Comment (mattklein123) [inline api/envoy/extensions/filters/http/decompressor/v3/decompressor.proto:54]: Can you add docs here to explain what this config does and what happens if it is not set? Presumably if empty it doesn't decompress in the given direction?

File: api/envoy/extensions/filters/http/decompressor/v3/decompressor.proto (hunk 4 above)
Comment (mattklein123) [reply api/envoy/extensions/filters/http/decompressor/v3/decompressor.proto:49]: FWIW it seemed odd to me that we don't allow this to be configured differently in each direction? Can you at least clarify this applies to both the request and response configuration depending on whether they are configured or not?

File: docs/root/configuration/http/http_filters/decompressor_filter.rst [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+By *default* decompression will be *skipped* when:
+
+- A request/response does NOT contain *content-encoding* header.
+- A request/response includes *content-encoding* header, but it does not contain the configured
+  decompressor's content-encoding.
+- A request/response contains a *cache-control* header whose value includes "no-transform".
+
+When decompression is *applied*:
+
+- The *content-length* is removed from headers.
```
Comment (mattklein123) [inline docs/root/configuration/http/http_filters/decompressor_filter.rst:45]: Is this going to break some uses cases? Is it a future work item that we might want to buffer the decompressed data and rewrite content-length?

File: source/extensions/compression/gzip/decompressor/config.cc [hunk 23]
```diff
@@ -9,11 +9,15 @@ namespace Decompressor {
 namespace {
 const uint32_t DefaultWindowBits = 12;
 const uint32_t DefaultChunkSize = 4096;
+// When summed to window bits, this tells zlib library to decompress gzip data per:
```
Comment (mattklein123) [inline source/extensions/compression/gzip/decompressor/config.cc:12]: nit: s/summed/logical ORd or something like that (even though sum is technically correct I think this is more clear)

File: api/envoy/extensions/filters/http/decompressor/v3/decompressor.proto (hunk 4 above)
Comment (junr03) [reply api/envoy/extensions/filters/http/decompressor/v3/decompressor.proto:49]: It's trivial to configure them independently, i.e allow different decompressors on each direction. If both you and @rojkov like the idea of allowing that, I can make the change.

File: api/envoy/extensions/filters/http/decompressor/v3/decompressor.proto (hunk 21 above)
Comment (junr03) [reply api/envoy/extensions/filters/http/decompressor/v3/decompressor.proto:54]: Yep, bears repeating here what I wrote in the runtime field.

The default is actually to decompress if empty. Do you prefer to _not_ decompress if empty?

Comment (mattklein123) [reply api/envoy/extensions/filters/http/decompressor/v3/decompressor.proto:54]: I think it's fine either way I would just clearly document it as it wasn't clear to me on first read through.

File: docs/root/configuration/http/http_filters/decompressor_filter.rst (hunk 22 above)
Comment (junr03) [reply docs/root/configuration/http/http_filters/decompressor_filter.rst:45]: Ah, sorry @mattklein123 , this context got dropped because github didn't allow me to move my original PR over.

We discussed here: https://github.com/junr03/envoy/pull/1#discussion_r411669576

And then I had this idea: https://github.com/envoyproxy/envoy/pull/11172#issuecomment-628949673

tl;dr: I originally implemented buffering here. But then decided to remove (and depend on the buffer filter) to keep this filter simpler. wdyt?

Comment (mattklein123) [reply docs/root/configuration/http/http_filters/decompressor_filter.rst:45]: Does the buffer filter reset content length? Anyway I think that's fine but I might make that clear in the documentation that the user might want to couple this with the buffer filter.

Comment (junr03) [reply docs/root/configuration/http/http_filters/decompressor_filter.rst:45]: > Does the buffer filter reset content length?

Yep, it sets it if not present, which made it ideal for this arrangement.

https://github.com/envoyproxy/envoy/blob/a18611c9a81a87b49470846eff1f62cc8507043b/source/extensions/filters/http/buffer/buffer_filter.cc#L100

Yeah I can add a section in the docs about it 👌

Comment (junr03) [reply docs/root/configuration/http/http_filters/decompressor_filter.rst:45]: Ultimately the issue is that the setup would still be a bit kludgy because of filter ordering, and the fact that the buffer filter can't be disabled independently for requests and responses. For Lyft's use case it'll be fine because we don't need to preserve content-length on responses.

So at the end of the day it might make sense to have this filter pay the price and do buffering itself. My proposal is to leave as is (with the new docs I added) to get this PR in, and then I can open an issue with a link to the commit that had buffering, it would just need to be updated a bit, and integration tests written (which shouldn't be hard at all). I can get to it as a low-pri.

Comment (mattklein123) [reply docs/root/configuration/http/http_filters/decompressor_filter.rst:45]: I think it's fine to document the limitation and we can revisit later.

File: api/envoy/extensions/filters/http/decompressor/v3/decompressor.proto (hunk 4 above)
Comment (junr03) [reply api/envoy/extensions/filters/http/decompressor/v3/decompressor.proto:49]: > LGTM modulo the comment about the compression config. Not sure if you are still planning on fixing that or not.

@mattklein123 do you mean this? Sorry, I was waiting for you and/or @rojkov to vote one way or another. But I guess both of you have already expressed interest in it, so I will go ahead and add.

Comment (rojkov) [reply api/envoy/extensions/filters/http/decompressor/v3/decompressor.proto:49]: IMHO it's a question of composability vs monolithic design. As a proponent of the former I'd prefer two sequential decompressor filters configured differently for different directions and only consider the latter if performance became an issue.

The same applies to buffering: coupled with the buffer filters we can get the same functionality with less code complexity. If the buffer filter is limited to requests only then it makes sense to extend it (and only it) - people might want to buffer responses in scenarios other than decompression (e.g. compression).

So, for now I vote for a comment on how to couple filters for various use cases instead of complicating the filter. Later we can reconsider.

Comment (mattklein123) [reply api/envoy/extensions/filters/http/decompressor/v3/decompressor.proto:49]: I'm fine either way as long as the current state is well documented.

Comment (junr03) [reply api/envoy/extensions/filters/http/decompressor/v3/decompressor.proto:49]: Alright, I had added documentation in the proto, but I have expanded with an example. The current proto would lend itself pretty easily to adding per direction config in the future is someone wants it.

lmk if you think this is clear @mattklein123 @rojkov https://github.com/envoyproxy/envoy/pull/11172/commits/858c279d82aab752ab59d14a5aebd2685e671115
